src/lug/lug.py

```diff
@@ -186,12 +191,113 @@ def unpatch_system_calls(patched_functions):
+            has_so_files = glob.glob(f"{adjusted_primary_location}/*.so") \
+                + glob.glob(f"{adjusted_primary_location}/**/*.so", recursive=True)
+            if has_so_files:
+                picklable = False
+
+        if not picklable:
+            try:
+                pip_names = packages_distributions[module.__name__]
+                for pip_name in pip_names:
+                    uncopyable_pip_names[pip_name] = module.__version__
+                uncopyable_packages.add(module.__name__)
+                # Make sure we know not to copy the child if present (e.g. lug and lug.run)
+                if child:
+                    uncopyable_packages.add(child.__name__)
+            except KeyError:
+                if not has_so_files:
+                    # Add to packages that need to be manually copied
+                    copyable_packages.add(primary_location)
+                else:
+                    raise ModuleNotFoundError(f"'{module.__name__}' distribution info not found on this machine.")
+            continue
+
+    return uncopyable_packages, uncopyable_pip_names, copyable_packages
+
+
+def env_requirements_to_string(pip_packages):
+    requirements_string = ""
+    for name, version in pip_packages.items():
+        requirements_string += f"{name}=={version} "
+    return requirements_string
+
+
+def create_python_script(func, args, kwargs, temp_input, user_docker, docker_shell_location, serialize_dependencies,
+                         internal_dir):
     output_uuid = uuid.uuid4()
     with open(temp_input.name, 'w') as fp:
+        copyable_packages = []
+        pip_packages_string = ''
+        links = []
         if serialize_dependencies:
             modules_to_register = get_modules_to_register(func)
-            for module in modules_to_register:
+            uncopyable_packages, uncopyable_pip_names, copyable_packages = find_module_transferability(
+                modules_to_register
+            )
+            pip_packages_string = env_requirements_to_string(uncopyable_pip_names)
+            pickle_packages = list(filter(
+                lambda mod: mod.__name__ not in copyable_packages and mod.__name__ not in uncopyable_packages,
+                modules_to_register,
+            ))
+            internal_dir_basename = os.path.basename(internal_dir)
+            for module_location_to_include in copyable_packages:
+                #  todo: handle conflicts for package directory names
+                local_path_with_internal_dir = os.path.join(
+                    internal_dir_basename,
+                    os.path.basename(module_location_to_include)
+                )
+                abs_path_with_internal_dir = os.path.join(internal_dir, os.path.basename(module_location_to_include))
+                if not os.path.exists(abs_path_with_internal_dir):
+                    shutil.copytree(module_location_to_include, abs_path_with_internal_dir)
```

Comment:
This copies all modules to a staging directory before being tarred with the other inputs. It's in a tmp directory, but it's not being explicitly released.

This shouldn't be a large amount of files – in testing, the modules that needed to be copied were usually small local references; the big modules usually required a pip install – but there isn't a guarantee.